go-runtime/encoding/encoding.go

```diff
@@ -198,3 +201,219 @@ func encodeString(v reflect.Value, w *bytes.Buffer) error {
 	w.WriteRune('"')
 	return nil
 }
+
+func Unmarshal(data []byte, v any) error {
+	rv := reflect.ValueOf(v)
+	if rv.Kind() != reflect.Ptr || rv.IsNil() {
+		return fmt.Errorf("unmarshal expects a non-nil pointer")
+	}
+	return decodeValue(data, rv.Elem())
+}
+
+func decodeValue(data []byte, v reflect.Value) error {
+	if !v.CanSet() {
+		return fmt.Errorf("cannot set value")
+	}
+
+	t := v.Type()
+	switch {
+	case v.CanAddr() && v.Addr().Type().Implements(jsonUnmarshaler):
+		v = v.Addr()
+		fallthrough
+
+	case t.Implements(jsonUnmarshaler):
+		dec := v.Interface().(json.Unmarshaler) //nolint:forcetypeassert
+		return dec.UnmarshalJSON(data)
+
+	case v.CanAddr() && v.Addr().Type().Implements(textUnmarshaler):
+		v = v.Addr()
+		fallthrough
+
+	case t.Implements(textUnmarshaler):
+		dec := v.Interface().(encoding.TextUnmarshaler) //nolint:forcetypeassert
+		var s string
+		if err := json.Unmarshal(data, &s); err != nil {
+			return err
+		}
+		return dec.UnmarshalText([]byte(s))
+	}
+
+	switch v.Kind() {
+	case reflect.Struct:
+		return decodeStruct(data, v)
+
+	case reflect.Ptr:
+		if v.IsNil() {
+			v.Set(reflect.New(v.Type().Elem()))
+		}
+		return decodeValue(data, v.Elem())
+
+	case reflect.Slice:
+		if v.Type().Elem().Kind() == reflect.Uint8 {
+			return decodeBytes(data, v)
+		}
+		return decodeSlice(data, v)
+
+	case reflect.Map:
+		return decodeMap(data, v)
+
+	case reflect.String:
+		return decodeString(data, v)
+
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return decodeInt(data, v)
+
+	case reflect.Float32, reflect.Float64:
+		return decodeFloat(data, v)
+
+	case reflect.Bool:
+		return decodeBool(data, v)
+
+	case reflect.Interface:
+		if v.Type().NumMethod() != 0 {
+			return fmt.Errorf("the only interface type supported is any, not %s", v.Type())
+		}
+
+		var anyInterface any
+		if err := json.Unmarshal(data, &anyInterface); err != nil {
+			return err
+		}
+		reflectedValue := reflect.ValueOf(anyInterface)
+		if !reflectedValue.Type().AssignableTo(v.Type()) {
+			return fmt.Errorf("cannot assign type %s to any interface", reflectedValue.Type())
+		}
+		v.Set(reflectedValue)
+		return nil
+
+	default:
+		return fmt.Errorf("unsupported type: %s", v.Type())
+	}
+}
+
+func decodeStruct(data []byte, v reflect.Value) error {
+	var jsonData map[string]json.RawMessage
+	if err := json.Unmarshal(data, &jsonData); err != nil {
+		return err
+	}
+
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		fieldType := v.Type().Field(i)
+		jsonKey := strcase.ToLowerCamel(fieldType.Name)
+		if _, ok := jsonData[jsonKey]; ok {
+			fieldTypeStr := fieldType.Type.String()
+			switch {
+			case fieldTypeStr == "*Unit" || fieldTypeStr == "Unit":
```

Comment:
This could be a direct comparison with `reflect.TypeOf(ftl.Unit{})` et al?

Reply:
this introduces a dependency cycle because `ftl` depends on encoding (it's used in `option.go`)